src/condor_schedd.V6/jobsets.cpp

```diff
@@ -59,248 +45,210 @@ JobSets::reconfig()
+				return false;
+			}
+			job.set_id = getOrCreateSet(setName, job.ownerinfo->name, new_ids);
+		}
 
-	IncrementLiveJobCounter(set->jobStatusAggregates, job.Universe(), old_status, -1);
-	IncrementLiveJobCounter(set->jobStatusAggregates, job.Universe(), new_status, 1);
+		JobQueueJobSet * set = getSet(job);
+		if (!set) {
+			// nothing to do, this job is not a member of a set
+			if (job.set_id > 0) {
+				// hmm. job has a setid, but that set does not exist
+			}
+			return false;
+		}
+		set->member_count += 1;
+		set->Assign(ATTR_REF_COUNT, set->member_count);
+		if (job.IsJob()) {
+			IncrementLiveJobCounter(set->jobStatusAggregates, job.Universe(), job.Status(), 1);
+		}
+		return true;
+	}
+	return false;
+}
 
-	if (IsDebugLevel(D_JOB) && old_status >= 0 ) {
-		ClassAd s;
-		set->jobStatusAggregates.publish(s, "Live");
-		if (set->setAd) {
-			MergeClassAds(&s, set->setAd, true);
+void JobSets::status_change(JobQueueJob & job, int new_status)
+{
+	if ( ! job.IsJob())
+		return;
+
+	JobQueueJobSet * jobset = getSet(job);
+	if (jobset) {
+		IncrementLiveJobCounter(jobset->jobStatusAggregates, job.Universe(), job.Status(), -1);
+		IncrementLiveJobCounter(jobset->jobStatusAggregates, job.Universe(), new_status, 1);
+
+		if (IsDebugCatAndVerbosity(D_JOB | D_VERBOSE)) {
+			ClassAd s;
+			jobset->jobStatusAggregates.publish(s, "Live");
+			MergeClassAds(&s, jobset, true);
+			std::string name;
+			if ( ! jobset->LookupString(ATTR_JOB_SET_NAME, name)) { name = std::to_string(jobset->Jobset()); }
+			dprintf(D_JOB | D_VERBOSE, "Updating JobSet %s (id=%u) after status change in job %d.%d to be:\n",
+				name.c_str(), jobset->Jobset(), job.jid.cluster, job.jid.proc);
+			dPrintAd(D_JOB, s);
 		}
-		dprintf(D_JOB, "Updating JobSet %s (id=%u) after status change in job %d.%d to be:\n",
-			set->name.c_str(), set->id, job.jid.cluster, job.jid.proc);
-		dPrintAd(D_JOB, s);
 	}
-	return true;
 }
 
 bool JobSets::removeJobFromSet(JobQueueJob & job)
 {
 	if (job.set_id <= 0) return true;
 
-	JobSet* jobset = getOrCreateSet(job);
-	ASSERT(jobset);
+	JobQueueJobSet* jobset = getSet(job);
+	if ( ! jobset) {
+		// if job is not in a set, we don't necessarily create one ;)
+		return true;
+	}
+
+	job.set_id = -1;
 
 	// update historical counters now to disk, as this job is leaving
-	if (jobset->setAd) {
-		std::string key;
-	
-		if (job.Status() == COMPLETED) {
-			key = ATTR_TOTAL_COMPLETED_JOBS;
-		}
-		if (job.Status() == REMOVED) {
-			key = ATTR_TOTAL_REMOVED_JOBS;
-		}
-		if (!key.empty()) {
-			if (job.Universe() == CONDOR_UNIVERSE_SCHEDULER) {
-				key = "Scheduler" + key;
-			}
+	// We need a separate count of jobs that are no longer in the set
+	// This is used to initialized the 
+	if (jobset) {
+		if (job.IsJob() && job.Status() >= REMOVED && job.Status() <= COMPLETED) {
+			static const char * const attrs[]{
+				ATTR_TOTAL_REMOVED_JOBS, "Scheduler" ATTR_TOTAL_REMOVED_JOBS,
+				ATTR_TOTAL_COMPLETED_JOBS, "Scheduler" ATTR_TOTAL_COMPLETED_JOBS,
```

Comment:
These attributes are in the wrong order. Unless we're planning to add aggregates for more statuses, this should be a few explicit if statements instead of computing an index into an array of attr names.
```suggestion
				ATTR_TOTAL_REMOVED_JOBS, ATTR_TOTAL_COMPLETED_JOBS,
				"Scheduler" ATTR_TOTAL_REMOVED_JOBS, "Scheduler" ATTR_TOTAL_COMPLETED_JOBS,
```

Reply:
On second thought, the expression should be fixed, not the order of the array. That will simplify adding aggregates for other job statuses.

Reply:
agreed.